[MLIR] Add pattern to bubble up tensor.extract_slice

ofri-frishman opened the pull request:
Add a pattern that bubbles up tensor.extract_slice through
tensor.expand_shape, and add a transform op to tensor dialect
to directly use this pattern.
This pattern enables tiling and fusing op chains which contain
tensor.expand_shape if added as a cleanup pattern of tile and fuse
utility.
Without this pattern that would not be possible, as
tensor.expand_shape does not implement the tiling interface.
In addition, registering this pattern as a cleanup pattern for
transform.structured.fuse.
The pattern was first implement in IREE project by
Quinn Dawkins and is being upstreamed.

---------

Co-authored-by: Quinn Dawkins <quinn.dawkins@gmail.com>

ofri-frishman committed:
[MLIR] Add pattern to bubble up tensor.extract_slice

Add a pattern that bubbles up tensor.extract_slice through
tensor.expand_shape, and add a transform op to tensor dialect
to directly use this pattern.
This pattern enables tiling and fusing op chains which contain
tensor.expand_shape if added as a cleanup pattern of tile and fuse
utility.
Without this pattern that would not be possible, as
tensor.expand_shape does not implement the tiling interface.
In addition, registering this pattern as a cleanup pattern for
transform.structured.fuse.
The pattern was first implement in IREE project by
Quinn Dawkins and is being upstreamed.

---------

Co-authored-by: Quinn Dawkins <quinn.dawkins@gmail.com>

mgehre-amd commented:
There are very similar patterns in mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp. How are they related? Should the new pattern also live in that file?

ofri-frishman commented:
I agree that the patterns in mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp are similar. The pattern I added is meant to swap between expand_shape and extract_slice which when used as a cleanup pattern for tile and fuse utility enables adding the expand_shape into a loop nest even though it does not implement the tiling interface.
I'm not familiar the the use of the other patterns, but don't think it is related to tiling.
I placed it in a separate file since I thought that other patterns that bubble up extract_slice through tensor ops could be placed there as well, so the emphasis is more on the extract_slice than on the expand_shape. But currently there is only one such pattern, so it could be placed together with the other reshape patterns.
But it would still require a separate populate function, since only this pattern should be added to the FuseOp cleanup patterns.
BTW there is another place with a similar pattern - in mlir/lib/Dialect/Linalg/Transforms/BubbleUpExtractSlice.cpp there is a pattern that bubbles up a extract_slice through a linalg op.
Given this, if you think that the best place for this pattern is in mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp I can move it to there

qedawkins commented on mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:
```diff
@@ -582,6 +582,7 @@ transform::FuseOp::apply(transform::TransformRewriter &rewriter,
     RewritePatternSet patterns(context);
     tensor::ExtractSliceOp::getCanonicalizationPatterns(patterns, context);
     tensor::populateMergeConsecutiveInsertExtractSlicePatterns(patterns);
+    tensor::populateBubbleUpExtractSliceOpPatterns(patterns);
```
I don't have a strong opinion on this transform op, but I don't know if I would bucket this pattern as "cleanup." If there are no objections this is fine with me.

ofri-frishman replied:
I agree that the pattern "unblocks" fusion of operations into a single tiling loop nest. 
I think that also other patterns used for cleanup assist with that. 
From what I understand, the concept of adding a possibility to cleanup between fusion steps was added precisely for this reason. As part of the thread in https://discourse.llvm.org/t/rfc-split-fusion-portions-of-the-tilinginterface-into-a-new-interface/81155 the idea to add such a step was proposed, and it was first implement in https://github.com/llvm/llvm-project/pull/109554. 
The lit tests added in that PR show cases where previously fusion would have got stuck, but after the change they do not. So, I thought it would make sense to add this pattern to the cleanup patterns as well.
But perhaps I misunderstood something about the nature and use of the cleanup patterns.
How otherwise would you suggest to add the ability to apply this pattern between fusion steps of `FuseOp`?

qedawkins replied:
Unless someone has a strong concern about this particular API detail, I just see `cleanup` as naming for "useful intermediate patterns," which this pattern I feel does apply to. Like noted, this is the intended API anyway, so the other solution would be to add another flag to `transform.structured.fuse`.

ofri-frishman replied:
I added a test where `apply_cleanup` is false and the bubble up does not happen.
I'm not sure if anyone has a case where they would require `apply_cleanup` not to be `True`, but since it is currently optional, I would not want to change that in this PR.
I think that perhaps the name "clean-up" isn't a good match for what these patterns do, a possible solution could be to change the name to something more fitting, but I would leave that to a separate PR.

banach-space replied:
We should consider removing this flag if it's not needed. Definitely a task fro a separate PR. You can leave a TODO there:
> TODO: It's not clear whether this flag is needed/used. Consider removing in the future.

Up to you :)

ofri-frishman replied:
I would prefer not to leave a TODO on this currently, as I haven't really made an effort to ask around and understand if anyone does have a use case where the flag is set to `false`.

mgehre-amd commented:
> I agree that the patterns in mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp are similar. The pattern I added is meant to swap between expand_shape and extract_slice which when used as a cleanup pattern for tile and fuse utility enables adding the expand_shape into a loop nest even though it does not implement the tiling interface. I'm not familiar the the use of the other patterns, but don't think it is related to tiling. I placed it in a separate file since I thought that other patterns that bubble up extract_slice through tensor ops could be placed there as well, so the emphasis is more on the extract_slice than on the expand_shape. But currently there is only one such pattern, so it could be placed together with the other reshape patterns. But it would still require a separate populate function, since only this pattern should be added to the FuseOp cleanup patterns. BTW there is another place with a similar pattern - in mlir/lib/Dialect/Linalg/Transforms/BubbleUpExtractSlice.cpp there is a pattern that bubbles up a extract_slice through a linalg op. Given this, if you think that the best place for this pattern is in mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp I can move it to there

Ok, make sense to me. Thanks for the analysis! Discoverability has been a big headache for me: There are many useful patterns, but its really hard to find whether a certain pattern already exists.
You can't probably do much about it in this PR.
Anyway, your patterns looks very useful, so thanks for the PR!

ofri-frishman commented:
Just a general question - currently in github I see that I am defined as a contributor to the llvm project and cannot add reviewers to review the PR and will not be able to land the PR once approved. Do you know who I can ask to give me the relevant permissions?

qedawkins commented:
Re: landing the PR. I don't know if the policy has changed, but typically for first time contributors a reviewer can land the PR once approved/ready.

qedawkins commented on mlir/lib/Dialect/Tensor/Transforms/BubbleUpExtractSlice.cpp:
```diff
@@ -0,0 +1,207 @@
+                                PatternRewriter &rewriter) const override {
+    auto expandOp = sliceOp.getSource().getDefiningOp<tensor::ExpandShapeOp>();
+    if (!expandOp) {
+      return failure();
```
```suggestion
      return rewriter.notifyMatchFailure(sliceOp,
                                         "slice source not produced by expand_shape");
```

ofri-frishman replied:
I amended the commit and updated according to your suggestion

qedawkins commented on mlir/lib/Dialect/Tensor/Transforms/BubbleUpExtractSlice.cpp:
```diff
@@ -0,0 +1,207 @@
+  auto newExpandShapeOp = rewriter.create<tensor::ExpandShapeOp>(
+      loc, resultType, newSliceOp, expandShapeOp.getReassociationIndices(),
+      sizes);
+  rewriter.replaceOp(sliceOp, newExpandShapeOp);
```
nit: Can use `replaceOpWithNewOp`

ofri-frishman replied:
I amended the commit and used `replaceOpWithNewOp` instead

qedawkins reviewed (commented):
Thanks for contributing this! I wrote most of the code here so it might be worth getting final approval from someone else. +1 to adding a separate testing op from `transform.structured.fuse`.

ofri-frishman commented:
> Thanks for contributing this! I wrote most of the code here so it might be worth getting final approval from someone else. +1 to adding a separate testing op from `transform.structured.fuse`.

I was actually wondering how best to go about that and wasn't aware of the option to add test ops, I was thinking about adding a test pass. You think the best way to have a test that isolates the pattern is via a test op (I guess you mean a transform op)? If so, any chance you could point me to an example of such an op?

ofri-frishman commented:
> Let me suggest a partial solution.
> 
> We should be able to identify relevant transformations by simply "grep"-ing for "interesting" Ops in the test directory. For this to work, we need to make sure that:
> 
> * we have good (meaningful + descriptive) test file names,
> * we have good (meaningful + descriptive) function names,
> * we test patterns in isolation and document what patterns are being excercised (e.g. [here](https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/vector-transfer-permutation-lowering.mlir#L5)).
> 
> As a stepping stone, @ofri-frishman, please make sure that the new pattern is tested in isolation. As for the right location, there seems to be 2 good candidates:
> 
> * "mlir/lib/Dialect/Linalg/Transforms/BubbleUpExtractSlice.cpp",
> * "mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp".
> 
> This pattern feels like "bubbling up" to me, though "swap" might be more fitting actually. Just to avoid bike-shedding, I suggest "BubbleUpExtractSlice.cpp", but lets note (in comments) that this is basically a "swap". We can re-visit later.

I amended the commit and added a separate transform op to apply the pattern directly + tests that isolate the pattern.
Moved the pattern to "mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp".
Added a comment about the pattern currently being a bubble up pattern but that there is a possibility to rework it into a swap pattern.

ofri-frishman commented:
Gentle reminder @banach-space @qedawkins

banach-space commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,74 @@
+// RUN: mlir-opt -split-input-file -transform-interpreter  %s | FileCheck %s
+
+// CHECK-LABEL:   func.func @bubble_up_extract_slice_through_expand_shape(
+// CHECK-SAME:                                                    %[[ARG0:.*]]: tensor<60xf32>) -> tensor<1x1x5xf32> {
```
[nit] Avoid such deep indentation.
[nit] Try to match MLIR and LIT variables (`%0` vs `%ARG0`).

ofri-frishman replied:
Updated

banach-space commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,74 @@
+// CHECK-LABEL:   func.func @bubble_up_extract_slice_through_expand_shape(
+// CHECK-SAME:                                                    %[[ARG0:.*]]: tensor<60xf32>) -> tensor<1x1x5xf32> {
+// CHECK:           %[[C1:.+]] = arith.constant 5 : index
+// CHECK:           %[[EXTRACT:.*]] = tensor.extract_slice %[[ARG0]]{{\[}}%[[C1]]] [5] [1] : tensor<60xf32> to tensor<5xf32>
```
```suggestion
// CHECK:           %[[EXTRACT:.*]] = tensor.extract_slice %[[ARG0]][%[[C1]]] [5] [1] : tensor<60xf32> to tensor<5xf32>
```

Similar suggestion elsewhere. IIRC, regex matching would be required if it was `[[` rather than `[`.

ofri-frishman replied:
updated here and in another location

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +213,178 @@ struct BubbleUpExpandThroughParallelCollapse
+        sliceOp.getSource().getDefiningOp<tensor::ExpandShapeOp>();
+    if (!expandShapeOp) {
+      return rewriter.notifyMatchFailure(
+          sliceOp, "slice source not produced by expand_shape");
```
[nit] Make a bit more verbose 
```suggestion
          sliceOp, "tensor.extract_slice source not produced by expand_shape");
```

ofri-frishman replied:
updated

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +213,178 @@ struct BubbleUpExpandThroughParallelCollapse
+          sliceOp, "slice source not produced by expand_shape");
+    }
+
+    if (!sliceOp.hasUnitStride()) {
+      return rewriter.notifyMatchFailure(sliceOp,
+                                         "unsupported: non-unit stride");
+    }
```
IIUC, the problem here is the lack of contiguity, right? It's worth adding that extra info.

ofri-frishman replied:
Yes. Added a few more words

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +213,178 @@ struct BubbleUpExpandThroughParallelCollapse
+        getMixedValues(expandShapeOp.getStaticOutputShape(),
+                       expandShapeOp.getOutputShape(), rewriter);
+
+    auto isZeroOffsetAndFullSize =
```
[nit] Spell out auto

ofri-frishman replied:
I was sure what you meant. I assumed you meant writing the type explicitly, so updated accordingly

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +213,178 @@ struct BubbleUpExpandThroughParallelCollapse
+                                         "unimplemented: rank reducing slice");
+    }
+
+    // Helper variables and function for accumulating the new offset and length
+    // values.
+    Location loc = expandShapeOp->getLoc();
+    AffineExpr d0, d1, d2;
+    bindDims(rewriter.getContext(), d0, d1, d2);
+    // Multiply two integers.
+    auto mul = [&](OpFoldResult v1, OpFoldResult v2) {
+      auto mulMap = AffineMap::get(2, 0, {d0 * d1});
+      return affine::makeComposedFoldedAffineApply(rewriter, loc, mulMap,
+                                                   {v1, v2});
+    };
```
IIUC, this needed for the actual "transformation". However, up until ~L321 this pattern is "analysing". I think that you can move this closer to where it's used.

[nit] I would also try to split "analysis" from "transformation" with some comment. Alternatively, move analysis to some helper hook (e.g. `checkPreconditionForBubbleUpExtractSlice`).

ofri-frishman replied:
I did a small refactor to move all the analysis code into a function named `checkPreconditionForBubbleUpExtractSlice`

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +213,178 @@ struct BubbleUpExpandThroughParallelCollapse
+        OpFoldResult fullSize = outputShape[indices[i]];
+        basis.push_back(fullSize);
+        delinOffsets.push_back(rewriter.getIndexAttr(0));
+        newSize = mul(newSize, fullSize);
```
I didn't see this size re-calculation in [bubble-up-extract-slice-op.mlir](https://github.com/llvm/llvm-project/pull/126898/files#diff-5b706598efea999ee91a9207f5bb81fb6e566754507af09c36456b8ee7f72252). It would be good to test it as well.

ofri-frishman replied:
Nice! I didn't notice that a test was missing for that. Added a relevant test.

banach-space commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
There's a lot of "negative" tests here (i.e. tests that show where the pattern does not apply). That's great, thank you! But perhaps there should be more positive tests as well? I can try to come up with some suggestions later.

ofri-frishman replied:
I added a few more positive tests to check some more cases

banach-space reviewed (commented):
Nice, thanks for the updates! And for the ping - this had slipped off my radar.

I’ve left some minor suggestions inline. I’ll take another look later.

ofri-frishman commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,74 @@
+    } : !transform.op<"func.func">
+    transform.yield
+  }
+}
```
Added

ofri-frishman commented:
> Nice, thanks for the updates! And for the ping - this had slipped off my radar.
> 
> I’ve left some minor suggestions inline. I’ll take another look later.

Thanks for the additional review, I've made changes according to your suggestions.

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+    for (const ReassociationIndices &indices :
+         expandShapeOp.getReassociationIndices()) {
+      OpFoldResult newSize = rewriter.getIndexAttr(1);
+      SmallVector<OpFoldResult> basis, delinOffsets;
```
What's `delinOffsets`?

banach-space commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+
+      int64_t i = 0;
+      int64_t e = indices.size();
+      // Offset = cumulative product of leading unit extracted dims.
```
This comment seems to apply specifically to "offsets", but this loop computes more than just offsets?

Perhaps document every loop so that it's clear what it's dealing with. Currently that's rather unclear.

ofri-frishman replied:
After reviewing the code again I came to the conclusion that given the checks done in the helper function, I could reduce the code here to a single loop.
So, I reduced it to a single loop and also made some variable name change and added a lot of documentation. Hopefully the code is much clearer now.

banach-space commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,115 @@
+// RUN: mlir-opt -split-input-file -transform-interpreter  %s | FileCheck %s
+
+// CHECK-LABEL:   func.func @bubble_up_extract_slice_through_expand_shape(
+// CHECK-SAME:                %[[VAL_0:.*]]: tensor<60xf32>) -> tensor<1x1x5xf32> {
+// CHECK:           %[[C1:.+]] = arith.constant 5 : index
+// CHECK:           %[[EXTRACT:.*]] = tensor.extract_slice %[[VAL_0]][%[[C1]]] [5] [1] : tensor<60xf32> to tensor<5xf32>
+// CHECK:           %[[EXPAND:.*]] = tensor.expand_shape %[[EXTRACT]] {{\[\[}}0, 1, 2]] output_shape [1, 1, 5] : tensor<5xf32> into tensor<1x1x5xf32>
+// CHECK:           return %[[EXPAND]] : tensor<1x1x5xf32>
+
+func.func @bubble_up_extract_slice_through_expand_shape(%0: tensor<60xf32>) -> tensor<1x1x5xf32> {
```
[nit] Matching MLIR and LIT variables names lower cognitive load, but it's just a nice-to-have :)
```suggestion
// CHECK-LABEL:   func.func @bubble_up_extract_slice_through_expand_shape(
// CHECK-SAME:                %[[SRC:.*]]: tensor<60xf32>) -> tensor<1x1x5xf32> {
// CHECK:           %[[C1:.+]] = arith.constant 5 : index
// CHECK:           %[[EXTRACT:.*]] = tensor.extract_slice %[[SRC]][%[[C1]]] [5] [1] : tensor<60xf32> to tensor<5xf32>
// CHECK:           %[[EXPAND:.*]] = tensor.expand_shape %[[EXTRACT]] {{\[\[}}0, 1, 2]] output_shape [1, 1, 5] : tensor<5xf32> into tensor<1x1x5xf32>
// CHECK:           return %[[EXPAND]] : tensor<1x1x5xf32>

func.func @bubble_up_extract_slice_through_expand_shape(%src: tensor<60xf32>) -> tensor<1x1x5xf32> {
```

ofri-frishman replied:
Thanks for clarifying, yeah makes sense. I Updated the tests accordingly

banach-space reviewed (commented):
Thanks for the updates!

The tests are incredibly helpful for understanding the code—very nicely formatted and consistent, too. Thank you!

I have a few minor points, but nothing major.

@qedawkins, what are your thoughts? Also, could you take a quick look at "transform-op-fuse.mlir" with your expert eye?

qedawkins commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+        getMixedValues(expandShapeOp.getStaticOutputShape(),
+                       expandShapeOp.getOutputShape(), rewriter);
+
+    std::function<bool(OpFoldResult, OpFoldResult, OpFoldResult)>
```
`auto` on lambdas is standard AIUI. I see that this came from a previous review comment so no need to go back to auto.

qedawkins reviewed (approved):
Tests look great, just a couple minor comments. Thanks for contributing this upstream!

(Make sure @banach-space's concerns are addressed before landing).

ofri-frishman commented on mlir/test/Dialect/Linalg/transform-op-fuse.mlir:
```diff
@@ -278,3 +278,141 @@ module attributes {transform.with_named_sequence} {
+
+
+
+
```
Fixed

ofri-frishman commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,115 @@
+  return %extract : tensor<1x1x5xf32>
+}
+
+
```
Fixed

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+              return llvm::succeeded(maybeEqual) && maybeEqual.value();
+            };
+
+    // First verify that this is a full slice of the expanded tensor.
```
1. I added a definition of a "full slice" in the documentation of `BubbleUpExpandShapeThroughExtractSlice`. There are a few examples of valid full slices using the ops below that. I also added a few examples here for contiguous and non contiguous slices.
2.  Updated the comment
3. as stated, there is `no_swap_expand_shape_with_extract_slice_non_contiguous`

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+    return success();
+  }
+
+  LogicalResult
+  checkPreconditionForBubbleUpExtractSlice(tensor::ExtractSliceOp sliceOp,
```
Added documentation

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,200 @@ struct BubbleUpExpandThroughParallelCollapse
+    for (const ReassociationIndices &indices :
+         expandShapeOp.getReassociationIndices()) {
+      OpFoldResult newSize = rewriter.getIndexAttr(1);
+      SmallVector<OpFoldResult> basis, delinOffsets;
```
Added a comment to explain the variables

banach-space replied:
Ha, now everything is clear!

[nit]  Reading your comment made me think: `basis` -> `reassocGroupSizes` and `delinOffsets` -> `reassocGroupOffsets`. These are rather mouthful, but then you could skip the comment. But honestly, feel free to ignore if you like the current names :)

banach-space reviewed (approved):
LGTM % some final nits

@ofri-frishman , you've done really great job refactoring, documenting and testing this, thank you 🙏🏻 Honestly, I find it so much to follow the code after your changes. 

I've left some minor nits/suggestions. Nothing major and hopefully not too much work.

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+/// taken out of the group could be defined as a single contiguous subrange
+/// within that range.
+/// If the transformation is not possible, or if the slice is rank reducing, the
+/// function returns failure.
```
updated based on your suggestion

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
````diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+/// function returns failure.
+///
+/// Example:
+/// ```
````
updated based on your suggestion

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+/// // The transformation is possible because each reassociation group has a
+/// // contiguous slice. (i.e., [2x4->2x4], [2x8->1x5], [4x2x4->1x1x4])
+/// // After the transformation:
+///
```
updated based on your suggestion

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+///
+/// // The transformation is possible because each reassociation group has a
+/// // contiguous slice. (i.e., [2x4->2x4], [2x8->1x5], [4x2x4->1x1x4])
+/// // After the transformation:
```
good suggestion, updated this

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
````diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+///     tensor<8x5x4xf32> to tensor<2x4x1x5x1x1x4xf32>
+/// ```
+///
+/// Note - this pattern could be reworked to be a swap pattern between
+/// `tensor.expand_shape` and `tensor.extract_slice`, but is currently
+/// implemented only as a bubble up pattern for `tensor.extract_slice`.
````
Maybe I don't understand correctly the distinction between a "swap" pattern and a "bubble up" pattern.
I though that one of the main distinctions between them is that a swap pattern can change the order of ops in both directions and bubble up only supports a single direction.
Currently, the pattern I added only supports moving a `tensor.extract_slice` that has a `tensor.expand_shape` as a producer to be before the `tensor.expand_shape` and doesn't support the other way around. So, I saw it as a bubble up pattern.
Am I missing something here in my understanding of the distinction between the types of patterns?

banach-space replied:
> Am I missing something here in my understanding of the distinction between the types of patterns?

Your taxonomy makes sense to me, but I haven't seen that documented anywhere 😅 Nor any API's/examples of that. But there's a lot of corners of MLIR that I haven't touched!

My concern is that there's a non-zero chance that we are discussing something that will not make sense to somebody else reading this comment and who didn't participate in this discussion. In fact, your tests in "transform-op-fuse.mlir" are already making references to "swapping": `@swap_expand_shape_with_extract_slice`. So, somebody might wonder, what's missing to make this a "swap" pattern?

I think that you can disambiguate this a bit by replacing "reworked" with "extended":
>  Note - this pattern could be extended ...

ofri-frishman replied:
I changed to working to "extended" and updated the func names in "transform-op-fuse.mlir" to use the term "bubble up" instead of "swap" to clarify.

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+    // of the tensor.expand_shape, so variables referring to the state before
+    // applying the pattern are named with the prefix "expanded", and ones
+    // referring to the state after applying the pattern are named with the
+    // prefix "collapsed".
```
Thanks, updated

ofri-frishman commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,217 @@ struct BubbleUpExpandThroughParallelCollapse
+      collapsedOffsets.push_back(collapsedOffset);
+      collapsedSizes.push_back(collapsedSize);
+
+      // Only unit stride supported.
```
updated

ofri-frishman commented on mlir/test/Dialect/Tensor/bubble-up-extract-slice-op.mlir:
```diff
@@ -0,0 +1,124 @@
+// CHECK:           %[[EXTRACT:.*]] = tensor.extract_slice %[[SRC]]{{\[}}%[[C0]]] {{\[}}%[[AFFINE_APPLY]]] [1] : tensor<60xf32> to tensor<?xf32>
+// CHECK:           %[[EXPAND:.*]] = tensor.expand_shape %[[EXTRACT]] {{\[\[}}0, 1, 2]] output_shape {{\[}}%[[SLICE_SIZE]], 5, 2] : tensor<?xf32> into tensor<?x5x2xf32>
+// CHECK:           return %[[EXPAND]] : tensor<?x5x2xf32>
+func.func @bubble_up_extract_slice_affine_apply_not_folded(%0: tensor<60xf32>, %slice_size : index) -> tensor<?x5x2xf32> {
```
updated

ofri-frishman commented:
Regarding the negative test for `BubbleUpExpandShapeThroughExtractSlice` within the tests for `transform.structured.fuse` - I understand your point, but I see this a bit differently.
I think that having such negative tests in `transform.structured.fuse`  help clarify what you could expect to work and not work for `transform.structured.fuse`. It isn't meant to get negative test coverage for the `BubbleUpExpandShapeThroughExtractSlice` as that is already achieved in the specific tests written directly for `BubbleUpExpandShapeThroughExtractSlice`.

ofri-frishman commented:
> LGTM % some final nits
> 
> @ofri-frishman , you've done really great job refactoring, documenting and testing this, thank you 🙏🏻 Honestly, I find it so much to follow the code after your changes.
> 
> I've left some minor nits/suggestions. Nothing major and hopefully not too much work.

Thanks for your time on this. I made some final updates based on your suggestions.

banach-space reviewed (approved):
LGTM, thanks!

@qedawkins has already approved, and I’m the only other reviewer, so this should be good to land.

Since it’s Friday, we can wait until Monday if you prefer—just in case the bots start complaining. 🙂 (They shouldn’t if pre-commit is 🟢, but you never know!)

Thanks for putting in all the effort on this - great work! 🙏🏻

ofri-frishman commented:
> LGTM, thanks!
> 
> @qedawkins has already approved, and I’m the only other reviewer, so this should be good to land.
> 
> Since it’s Friday, we can wait until Monday if you prefer—just in case the bots start complaining. 🙂 (They shouldn’t if pre-commit is 🟢, but you never know!)
> 
> Thanks for putting in all the effort on this - great work! 🙏🏻

@banach-space I made a few final tweaks. Since I don't have write permissions yet, could you land this PR for me?

ofri-frishman commented:
@mgehre-amd @banach-space @qedawkins thank you very much for the time spent on reviewing this PR, it is greatly appreciated.
This is my first meaningful contribution to LLVM repository, and your comments and suggestions are instrumental in making sure it is a high quality contribution.

banach-space commented:
A bit of book-keeping before I land this.

From you summary:
> The pattren was first implement in IREE project by Quinn Dawkins and is being upstreamed.

1. You may want to add Quinn as a co-author.
2. There's a typo in your summary ("pattren" -> "pattern")

Also, when I try to merge it, GitHub adds this note:
> Co-authored-by: Ofri Frishman <ofri.frishman@mobileye.com>

It also "prints" this note:

> This commit will be authored by 32477595+ofri-frishman@users.noreply.github.com.

Please make sure that you have your e-mail address correctly set-up:
* https://llvm.org/docs/DeveloperPolicy.html#github-email-address

I suspect that that's the reason for all the extra noise from GitHub.

Thanks!

ofri-frishman commented:
> A bit of book-keeping before I land this.
> 
> From you summary:
> 
> > The pattren was first implement in IREE project by Quinn Dawkins and is being upstreamed.
> 
> 1. You may want to add Quinn as a co-author.
> 2. There's a typo in your summary ("pattren" -> "pattern")
> 
> Also, when I try to merge it, GitHub adds this note:
> 
> > Co-authored-by: Ofri Frishman [ofri.frishman@mobileye.com](mailto:ofri.frishman@mobileye.com)
> 
> It also "prints" this note:
> 
> > This commit will be authored by [32477595+ofri-frishman@users.noreply.github.com](mailto:32477595+ofri-frishman@users.noreply.github.com).
> 
> Please make sure that you have your e-mail address correctly set-up:
> 
> * https://llvm.org/docs/DeveloperPolicy.html#github-email-address
> 
> I suspect that that's the reason for all the extra noise from GitHub.
> 
> Thanks!

1. Added Quinn as a co-author. I wasn't aware of the convention of adding co-authors as a separate part after the commit message, thanks.
2. Fixed the typo.
I have a bit of a mess between my private and work emails. I changed the email in the commit message to my private one which is connected to the GitHub account, and updated the email settings in GitHub. Could you see if now GitHub doesn't add extra lines?

banach-space commented:
> Added Quinn as a co-author.

Thanks!

Just a heads-up: you’ve updated the commit message instead of the PR summary. When I **press "Squash and merge"** on your behalf, GitHub will use the **PR summary** rather than the commit message (*). This is covered in the LLVM documentation (https://llvm.org/docs/GitHub.html#github-reviews): 
> Use the button Squash and merge in GitHub’s web interface. If you do this, remember to review the commit message when prompted.

We should probably add a clear note about using the PR summary - there’s no way for a new contributor to see this info in the UI 😅 (I only know it because I’ve done it _a few_ times).  

Additionally, I noticed you’ve been **force-pushing** instead of using **fix-up commits**. LLVM's policy recommends using fix-up commits instead of force-pushing:  

* https://llvm.org/docs/GitHub.html#updating-pull-requests 
> When updating a pull request, you should push additional “fix-up” commits to your branch instead of force pushing.

---

### **On Co-Authors in Commit Messages**
> I wasn't aware of the convention of adding co-authors as a separate part after the commit message, thanks.

I had assumed this was already covered by our **code-review policy**, but it seems like it isn’t:  
* https://discourse.llvm.org/t/code-review-process-when-upstreaming-patches

Regardless, it’s a good practice to follow - we should ensure **all contributions are properly attributed** (not to take anything away from your fantastic effort here!).  

—Andrzej  

---

(*) Otherwise, in the case of multiple commits, GitHub wouldn’t know which commit message to use (all of them?).

ofri-frishman commented:
Thanks for all the important info.
- I updated the message at the top to match the commit message. I hope that is the PR summary that you mentioned.
- I wasn't really sure when I started work on this if I should force merge or create additional commits, since I didn't know if squashing commits was possible. I'll know to use that for next time.
Do you see any other updates that are required?

banach-space commented:
> * I wasn't really sure when I started work on this if I should force merge or create additional commits, since I didn't know if squashing commits was possible. 

TBH, finding that info is not the easiest thing ATM. I know that some folks are working hard to improve that, but most of these improvements are hobby projects and these tend to take time. 

That said, you are a new contributor and it's for the more established contributors to mentor you :) So, mea culpa for not pointing these things out earlier!

I am landing this shortly, thanks for bearing with me 🙏🏻

hanhanW commented on mlir/lib/Dialect/Tensor/Transforms/ReshapePatterns.cpp:
```diff
@@ -210,6 +214,214 @@ struct BubbleUpExpandThroughParallelCollapse
+/// Note - this pattern could be extended to be a swap pattern between
+/// `tensor.expand_shape` and `tensor.extract_slice`, but is currently
+/// implemented only as a bubble up pattern for `tensor.extract_slice`.
+struct BubbleUpExpandShapeThroughExtractSlice
```
I'm coming from https://github.com/llvm/llvm-project/pull/153675, which performs the refactoring.

I'm sorry that I did not ask this question in the first place. May I ask why you name it as `BubbleUpExpandShapeThroughExtractSlice`, but the (first) test name is `bubble_up_extract_slice_through_expand_shape`? Should it be `BubbleUpExtractSliceThroughExpandShape`?

ofri-frishman replied:
You are asking about the order of operations in the name and not about camel case vs. "_" right?
If that is the case then it seems that the name is incorrect and should be `BubbleUpExtractSliceThroughExpandShape` as you suggested.

hanhanW replied:
Yes, I also checked other patterns in the upstream, but some of them were created by me. I'm not a native speaker, so I'm not sure if I made a mistake or not. Thanks for checking, I'll prepare a PR after we land the refactoring change.
